aten/src/ATen/native/mkl/SparseCsrLinearAlgebra.cpp

```diff
@@ -0,0 +1,217 @@
+#include <ATen/ATen.h>
+#include <ATen/SparseCsrTensorUtils.h>
+
+// Don't compile with MKL for MSVC/macos since linking the sparse MKL routines
+// needs some build fixes.
+// https://github.com/pytorch/pytorch/pull/50937#issuecomment-778732740
+// Macros source: https://web.archive.org/web/20191012035921/http://nadeausoftware.com/articles/2012/01/c_c_tip_how_use_compiler_predefined_macros_detect_operating_system
+#if !AT_MKL_ENABLED() || defined(_MSC_VER) || defined(__APPLE__) || defined(__MACH__)
+
+namespace at {
+namespace native {
+using namespace at::sparse;
+Tensor& _sparse_mm_mkl_(
+    Tensor& self,
+    const SparseTensor& sparse_,
+    const Tensor& dense,
+    const Tensor& t,
+    const Scalar& alpha,
+    const Scalar& beta) {
+#if _MSC_VER
+  AT_ERROR("sparse_mm_mkl: MKL support is disabled on Windows");
+#elif __APPLE__ || __MACH__
+  AT_ERROR("sparse_mm_mkl: MKL support is disabled on macos/iOS.");
+#else
+  AT_ERROR("sparse_mm_mkl: ATen not compiled with MKL support");
+#endif
+}
+} // namespace native
+} // namespace at
+
+#else // AT_MKL_ENABLED
+
+#include <ATen/mkl/Descriptors.h>
+#include <ATen/mkl/Exceptions.h>
+#include <ATen/mkl/Limits.h>
+#include <mkl.h>
+#include <mkl_spblas.h>
+
+#include <ATen/Dispatch.h>
+#include <ATen/ExpandUtils.h>
+#include <ATen/SparseCsrTensorImpl.h>
+
+#ifdef MKL_ILP64
+#define TORCH_INT_TYPE at::kLong
+#else
+#define TORCH_INT_TYPE at::kInt
+#endif
+
+namespace at {
+namespace native {
+using namespace at::sparse;
+
+static inline void sparse_mm_mkl_impl(
+    float* res,
+    MKL_INT* col_indices,
+    MKL_INT* crow_indices,
+    float* values,
+    float* dense,
+    float* t,
+    float alpha,
+    float beta,
+    MKL_INT nrows,
+    MKL_INT ncols,
+    MKL_INT dense_ncols) {
+  sparse_matrix_t A = 0;
+  matrix_descr desc;
+  desc.type = SPARSE_MATRIX_TYPE_GENERAL;
+  int retval = mkl_sparse_s_create_csr(
+      &A,
+      SPARSE_INDEX_BASE_ZERO,
+      nrows,
+      ncols,
+      crow_indices,
+      crow_indices + 1,
+      col_indices,
+      values);
+  TORCH_CHECK(
+      retval == 0, "mkl_sparse_s_create_csr failed with error code: ", retval);
+
+  mkl_sparse_s_mm(
+      SPARSE_OPERATION_NON_TRANSPOSE,
+      alpha,
+      A,
+      desc,
+      SPARSE_LAYOUT_ROW_MAJOR,
+      dense,
+      dense_ncols,
+      dense_ncols,
+      beta,
+      res,
+      dense_ncols);
+  mkl_sparse_destroy(A);
+}
+
+static inline void sparse_mm_mkl_impl(
```

Comment:
There's quite a bit of repetition here and with the float version above. An alternate way of structuring this code which could reduce duplication is to templatize the equivalent of `mkl_sparse_s_mm` and the CSR constructor itself, and then have a generic template function which calls into these two. We do this pattern for other BLAS calls because the signatures are usually very regular over types, and it seems like this should apply here too.

Reply:
I've made changes for RAII, however its not clear to me how that would be different in terms of LOC from overloading given that these functions are called only once.